### PR TITLE
Get role for switch and controllerId from configuration file.

### PR DESCRIPTION
**Pull request by tulioalberton**
[junit] Tests run: 27, Failures: 0, Errors: 16, Skipped: 0, Time elapsed: 0,213 sec                                                                                             
    [junit] TEST net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer10Test FAILED                                                                                    
    [junit] Running net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer13Test                                                                                        
    [junit] Tests run: 29, Failures: 0, Errors: 16, Skipped: 0, Time elapsed: 0,374 sec                                                                                             
    [junit] TEST net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer13Test FAILED   

Some problems with the tests.

Could you help me with the tests?

Regards

Tulio Ribeiro

### CODE REVIEW

**Comment by rizard** [comment]
@tulioalberton Odd -- TravisCI says the tests passed. Do they fail for you?

**Comment by tulioalberton** [comment]
Hi, yes the handshake fails for me.

Follow the errors:

[junit] Running 
net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer10Test
     [junit] Tests run: 27, Failures: 0,*Errors: 16*, Skipped: 0, Time 
elapsed: 0,121 sec
     [junit] TEST 
net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer10Test*FAILED*
     [junit] Running 
net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer13Test
     [junit] Tests run: 29, Failures: 0,*Errors: 16*, Skipped: 0, Time 
elapsed: 0,129 sec
     [junit] TEST 
net.floodlightcontroller.core.internal.OFSwitchHandshakeHandlerVer13Test*FAILED

Tomorrow I'll get the main branch of floodlight and test again.


Regards

*

Att,

Tulio Ribeiro

**Comment by rizard** [comment]
@tulioalberton did you intend to close this?

**Comment by tulioalberton** [comment]
Sorry, I'm new on github

Att,

Tulio Ribeiro

**Comment by tulioalberton** [comment]
Ryan, now the tests for HandShake with role from floodlightdefault.properties it is ok!
The tests for Sync when I do from Eclipse it is ok also!

Regards.

File: src/main/java/net/floodlightcontroller/core/internal/OFChannelHandler.java [hunk 1]
```diff
@@ -687,7 +688,7 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
 
 	@Override
 	public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-		log.debug("channelIdle on OFChannelHandler {}", String.format("%08x", System.identityHashCode(this)));
+		//log.debug("channelIdle on OFChannelHandler {}", String.format("%08x", System.identityHashCode(this)));
```
Comment (rizard) [inline src/main/java/net/floodlightcontroller/core/internal/OFChannelHandler.java]: @tulioalberton let's leave this debug message in there. If you want to change the log level to not show this, you can adjust the logback.xml file.

File: src/main/java/org/sdnplatform/sync/ISyncService.java [hunk 2]
```diff
@@ -133,4 +137,11 @@ public void registerPersistentStore(String storeName, Scope scope)
                        IInconsistencyResolver<Versioned<V>> resolver)
                                throws UnknownStoreException;
 
+
+    
+
```
Comment (rizard) [inline src/main/java/org/sdnplatform/sync/ISyncService.java]: @tulioalberton could you add some Javadoc style comments to describe what these new ISyncService functions do and return?

File: src/main/java/org/sdnplatform/sync/internal/SyncManager.java [hunk 3]
```diff
@@ -812,4 +815,12 @@ private SyncMessage getMessage(SynchronizingStorageEngine store) {
 			return bsm;
 		}
 	}
+
+	@Override
```
Comment (rizard) [inline src/main/java/org/sdnplatform/sync/internal/SyncManager.java]: @tulioalberton Ah, this is what you were talking about earlier with the RPCService and needing a listener, right? Right now, if I understand correctly, you're polling for the connections periodically and would like to have a way for the RPCService to notify you upon a connection add/remove event.

File: src/main/resources/logback-test.xml [hunk 4]
```diff
@@ -15,10 +15,19 @@
   <logger name="LogService" level="DEBUG"></logger> <!-- Restlet access logging -->
   <logger name="net.floodlightcontroller" level="INFO"/>
   <logger name="org.sdnplatform" level="INFO"></logger>
+  <logger name="org.sdnplatform.sync.internal.config.SyncStoreCCProvider" level="TRACE"></logger>
```
Comment (rizard) [inline src/main/resources/logback-test.xml]: @tulioalberton In commits to master, let's reduce all logs to INFO to not clutter stdout for average users :-)

**Comment by rizard** [comment]
@tulioalberton all TravisCI tests pass too, which is great! As indicated in my comments above, if you can address the following issues, I'll merge with master:
(1) add back the debug log in OFChannelHandler
(2) add some comments describing usage of the two new functions you added to ISyncService
(3) adjust all packages in logback.xml to INFO messages

and one additional request:
(4) rename or change floodlightdefault.properties2 to something more descriptive

Thanks so much! This is a great step forward for using the sync module! Please follow up with any updates to your code as you progress!

**Comment by tulioalberton** [comment]
I'll do your comments : )

File: src/main/java/org/sdnplatform/sync/ISyncService.java [hunk 5]
```diff
@@ -133,4 +133,18 @@ public void registerPersistentStore(String storeName, Scope scope)
...
+    
+   
+    /**
+     * Listener to RPC connections, used in simple Fault tolerance module,
+     * or other modules that need cluster monitor nodes (connect and disconnect).
+     * The listener is dispatched at connected or disconnected events.
+     * The listener inform the node id of connected or disconnected nodes on cluster.
+     * @param listener
+     */
+	public void addRPCListener(IRPCListener listener);
```
Comment (rizard) [inline src/main/java/org/sdnplatform/sync/ISyncService.java]: @tulioalberton very nice!

File: src/main/java/org/sdnplatform/sync/internal/SyncManager.java [hunk 6]
```diff
@@ -812,4 +814,29 @@ private SyncMessage getMessage(SynchronizingStorageEngine store) {
...
 		}
 	}
+
+	
+	/**
+	 * Add a listener to RPC connections on cluster.
+	 * The listener is dispatched at connect or disconnect events. 
+	 * Can be used to monitor connections or disconnections on cluster configs.
+	 */
+	@Override
```
Comment (rizard) [inline src/main/java/org/sdnplatform/sync/internal/SyncManager.java]: @tulioalberton love the listener! great work!

File: src/main/java/org/sdnplatform/sync/internal/rpc/IRPCListener.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package org.sdnplatform.sync.internal.rpc;
+
+public interface IRPCListener {
```
Comment (rizard) [inline src/main/java/org/sdnplatform/sync/internal/rpc/IRPCListener.java]: @tulioalberton again, very nice!

File: src/main/java/org/sdnplatform/sync/internal/rpc/RPCService.java [hunk 8]
```diff
@@ -502,6 +521,16 @@ protected void nodeConnected(short nodeId, Channel channel) {
             }
             c.nodeChannel = channel;
             c.state = NodeConnectionState.CONNECTED;
+
```
Comment (rizard) [inline src/main/java/org/sdnplatform/sync/internal/rpc/RPCService.java]: @tulioalberton seems like a very straightforward and clean fix

**Comment by rizard** [comment]
@tulioalberton this looks great! If you're ready and don't have anymore commits to add right now, let me know and I'll merge it

**Comment by tulioalberton** [comment]
Thanks @rizard, I have not more commits for now. 

But I'm planning to improve the simple.FT class, but I think you can merge and close the pull request. 
After improvements I'll do another pull request. 
Could be?  

Thank you for the comments...
Regards : )

**Comment by rizard** [comment]
@tulioalberton sorry for the delay! I'll merge now. Thanks for the support and contribution! BTW, if you'd like to contribute a brief tutorial/how-to for using the sync module with multiple controllers, please feel free. I'd be happy to put it on the wiki for you and give you full credit for it

**Comment by tulioalberton** [comment]
Hi Ryan


Where do I write the tutorial?
There is a special place for it?

Regards

Att,

Tulio Ribeiro

**Comment by rizard** [comment]
We like to keep all tutorials here:

https://floodlight.atlassian.net/wiki/display/floodlightcontroller/Tutorials

Feel free to create an account. Or, alternatively, you can write it however
you want and I can add it to the wiki when you're ready.

Thanks for helping out!

Ryan Izard
PhD Candidate, Research/Teaching Assistant
ECE Department, Clemson University
rizard@g.clemson.edu
---------------------------------------------------
Big Switch Networks
ryan.izard@bigswitch.com

**Pull request merged**